File: scripts/lib/resolve_credential_sources.py
#!/usr/bin/env python3
"""Resolve host-side credential inputs into concrete files before rendering."""

from __future__ import annotations

import argparse
import json
import os
from pathlib import Path
import shutil
import sys
import tempfile

SCRIPT_DIR = Path(__file__).resolve().parent
if str(SCRIPT_DIR) not in sys.path:
    sys.path.insert(0, str(SCRIPT_DIR))

from policy_bundle import (  # noqa: E402
    AGENT_SCOPED_CREDENTIAL_KEYS,
    SHARED_CREDENTIAL_KEYS,
    SUPPORTED_AGENTS,
    SUPPORTED_MODES,
    die,
    logical_policy_path,
    load_policy_bundle,
    require_secret_file,
    rebase_policy_fragment,
    render_policy_toml,
    selected_for,
    validate_allowed_keys,
    validate_source_path,
)


RESOLVER_ALLOWED_KEYS = {"source", "resolver", "materialization", "providers", "modes"}
SUPPORTED_MATERIALIZATION = {"ephemeral", "persistent"}
ALLOWED_RESOLVERS = {
    "claude_auth": {"claude-macos-keychain"},
}
TEST_CLAUDE_EXPORT_ENV = "WORKCELL_TEST_CLAUDE_KEYCHAIN_EXPORT_FILE"


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser()
    parser.add_argument("--policy", required=True)
    parser.add_argument("--agent", required=True, choices=sorted(SUPPORTED_AGENTS))
    parser.add_argument("--mode", required=True, choices=sorted(SUPPORTED_MODES))
    parser.add_argument(
        "--resolution-mode",
        required=True,
        choices=("launch", "metadata"),
    )
    parser.add_argument("--output-policy", required=True)
    parser.add_argument("--output-metadata", required=True)
    parser.add_argument("--output-root", required=True)
    return parser.parse_args()


def materialize_file(source: Path, destination: Path) -> None:
    destination.parent.mkdir(parents=True, exist_ok=True)
    with tempfile.NamedTemporaryFile(dir=destination.parent, delete=False) as handle:
        temp_path = Path(handle.name)
    try:
        shutil.copyfile(source, temp_path)
        temp_path.chmod(0o600)
        temp_path.replace(destination)
        destination.chmod(0o600)
    finally:
        try:
            temp_path.unlink()
        except FileNotFoundError:
            pass


def write_placeholder(destination: Path, resolver: str) -> None:
    destination.parent.mkdir(parents=True, exist_ok=True)
    with tempfile.NamedTemporaryFile(
        dir=destination.parent,
        delete=False,
        mode="w",
        encoding="utf-8",
    ) as handle:
        temp_path = Path(handle.name)
        handle.write(
            json.dumps(
                {
                    "workcell": "metadata-only",
                    "resolver": resolver,
                },
                sort_keys=True,
            )
            + "\n"
        )
    try:
        temp_path.chmod(0o600)
        temp_path.replace(destination)
        destination.chmod(0o600)
    finally:
        try:
            temp_path.unlink()
        except FileNotFoundError:
            pass


def atomic_write_text(destination: Path, content: str) -> None:
    destination.parent.mkdir(parents=True, exist_ok=True)
    with tempfile.NamedTemporaryFile(
        dir=destination.parent,
        delete=False,
        mode="w",
        encoding="utf-8",
    ) as handle:
        temp_path = Path(handle.name)
        handle.write(content)
    try:
        temp_path.chmod(0o600)
        temp_path.replace(destination)
        destination.chmod(0o600)
    finally:
        try:
            temp_path.unlink()
        except FileNotFoundError:
            pass


def resolve_claude_macos_keychain(
    destination: Path,
    resolution_mode: str,
) -> str:
    test_export_path = os.environ.get(TEST_CLAUDE_EXPORT_ENV, "")
    if test_export_path:
        source = require_secret_file(
            validate_source_path(
                test_export_path,
                TEST_CLAUDE_EXPORT_ENV,
                Path.cwd(),
            ),
            TEST_CLAUDE_EXPORT_ENV,
        )
        materialize_file(source, destination)
        return "resolved"
    if resolution_mode == "metadata":
        write_placeholder(destination, "claude-macos-keychain")
        return "configured-only"
    die(
        "Claude macOS login reuse is configured but no supported export path is "
        "available. Use claude_api_key or remove credentials.claude_auth."
    )


def resolve_credential(
    key: str,
    resolver_name: str,
    destination: Path,
    resolution_mode: str,
) -> str:
    if key == "claude_auth" and resolver_name == "claude-macos-keychain":
        return resolve_claude_macos_keychain(destination, resolution_mode)
    die(f"unsupported credential resolver: {resolver_name}")


def main() -> int:
    args = parse_args()
    policy_path = Path(args.policy).expanduser().resolve()
    output_policy_path = Path(args.output_policy).expanduser().resolve()
    output_metadata_path = Path(args.output_metadata).expanduser().resolve()
    output_root = Path(args.output_root).expanduser().resolve()

    policy, policy_sources = load_policy_bundle(policy_path)
    policy = rebase_policy_fragment(policy, policy_path.parent)
    credentials = policy.get("credentials", {})
    if credentials is None:
        credentials = {}
    if not isinstance(credentials, dict):
        die("credentials must be a TOML table")

    relevant_keys = SHARED_CREDENTIAL_KEYS | AGENT_SCOPED_CREDENTIAL_KEYS.get(args.agent, set())
    metadata = {
        "policy_entrypoint": logical_policy_path(policy_path, policy_path.parent),
        "policy_sources": policy_sources,
        "credential_input_kinds": {},
        "credential_resolvers": {},
        "credential_materialization": {},
        "credential_resolution_states": {},
    }

    for key in sorted(relevant_keys):
        raw = credentials.get(key)
        if raw is None:
            continue
        if not isinstance(raw, dict):
            metadata["credential_input_kinds"][key] = "source"
            metadata["credential_resolution_states"][key] = "source"
            continue

        validate_allowed_keys(raw, RESOLVER_ALLOWED_KEYS, f"credentials.{key}")
        resolver_name = raw.get("resolver")
        providers = raw.get("providers")
        modes = raw.get("modes")
        if not selected_for(
            providers,
            args.agent,
            f"credentials.{key}.providers",
            SUPPORTED_AGENTS,
        ):
            if resolver_name is not None:
                credentials.pop(key, None)
            continue
        if not selected_for(
            modes,
            args.mode,
            f"credentials.{key}.modes",
            SUPPORTED_MODES,
        ):
            if resolver_name is not None:
                credentials.pop(key, None)
            continue

        source_raw = raw.get("source")
        if source_raw is not None and resolver_name is not None:
            die(f"credentials.{key} must not declare both source and resolver")
        if source_raw is None and resolver_name is None:
            die(f"credentials.{key} must declare source or resolver")
        if resolver_name is None:
            metadata["credential_input_kinds"][key] = "source"
            metadata["credential_resolution_states"][key] = "source"
            continue
        if key not in ALLOWED_RESOLVERS or resolver_name not in ALLOWED_RESOLVERS[key]:
            die(f"credentials.{key}.resolver is unsupported: {resolver_name}")
        materialization = raw.get("materialization", "ephemeral")
        if materialization not in SUPPORTED_MATERIALIZATION:
            die(
                f"credentials.{key}.materialization must be one of: "
                f"{', '.join(sorted(SUPPORTED_MATERIALIZATION))}"
            )
        if materialization != "ephemeral":
            die(
                f"credentials.{key}.materialization must stay ephemeral for resolver-backed auth"
            )

        destination = output_root / "resolved" / "credentials" / f"{key}.json"
        state = resolve_credential(key, resolver_name, destination, args.resolution_mode)
        rewritten = dict(raw)
        rewritten.pop("resolver", None)
        rewritten.pop("materialization", None)
        rewritten["source"] = str(destination)
        credentials[key] = rewritten

        metadata["credential_input_kinds"][key] = "resolver"
        metadata["credential_resolvers"][key] = resolver_name
        metadata["credential_materialization"][key] = materialization
        metadata["credential_resolution_states"][key] = state

    atomic_write_text(output_policy_path, render_policy_toml(policy))
    atomic_write_text(
        output_metadata_path,
        json.dumps(metadata, sort_keys=True, indent=2) + "\n",
    )
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: internal/authresolve/resolve_credential_sources.go
// SPDX-License-Identifier: Apache-2.0
// Copyright 2026 Omkhar Arasaratnam

package authresolve

import (
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"os"
	"path/filepath"
	"runtime"
	"slices"
	"strings"

	"github.com/omkhar/workcell/internal/adapters"
	"github.com/omkhar/workcell/internal/cliexit"
	"github.com/omkhar/workcell/internal/injectionpolicy"
	"github.com/omkhar/workcell/internal/providerid"
	"github.com/omkhar/workcell/internal/rootio"
)

var (
	supportedAgents = providerid.AllProviderSet()
	supportedModes  = map[string]struct{}{
		"strict":      {},
		"development": {},
		"build":       {},
		"breakglass":  {},
	}
	supportedMaterialization = map[string]struct{}{
		"ephemeral":  {},
		"persistent": {},
	}
	allowedResolvers = map[string]map[string]struct{}{
		"codex_auth": {
			"codex-home-auth-file": {},
		},
		"claude_auth": {
			"claude-macos-keychain": {},
		},
	}
	sharedCredentialKeys      = adapters.SharedCredentialKeys()
	agentScopedCredentialKeys = adapters.AgentScopedCredentialKeysForProviders(providerid.AllProviders)
	allCredentialKeys         = credentialKeyUnion(agentScopedCredentialKeys, sharedCredentialKeys)
	rootPolicyKeys            = map[string]struct{}{
		"version":     {},
		"includes":    {},
		"documents":   {},
		"ssh":         {},
		"copies":      {},
		"credentials": {},
		"network":     {},
	}
	documentKeys = providerid.DocumentKeySet()
	// CredentialEntryKeys is the set of keys accepted in a
	// `[credentials.<name>]` table before credential resolution — the
	// resolver form.  It is the exact set validateAllowedKeys enforces in
	// run() for the table-shaped credential entry, and it is exported read-only
	// so the injection-policy schema drift check
	// (internal/injection/schema_doc_drift_test.go) can cross-check the
	// documented resolver-form key table against this parser set.
	CredentialEntryKeys = map[string]struct{}{
		"source":          {},
		"resolver":        {},
		"materialization": {},
		"providers":       {},
		"modes":           {},
	}
	systemSymlinkAllowlist = func() map[string]struct{} {
		if runtime.GOOS == "darwin" {
			return map[string]struct{}{
				"/var": {},
				"/tmp": {},
			}
		}
		return map[string]struct{}{}
	}()
)

// validateNetworkPolicy fails closed on an invalid [network] table before any
// credential is resolved, matching the render-time check in
// internal/injection.renderNetwork: only allow_endpoints/deny_endpoints are
// accepted (a mode-shaped key such as network_policy is rejected), and every
// endpoint must match the shared enforcement-helper grammar.
func validateNetworkPolicy(policy map[string]any) error {
	raw, ok := policy["network"]
	if !ok || raw == nil {
		return nil
	}
	network, ok := raw.(map[string]any)
	if !ok {
		return errors.New("network must be a TOML table")
	}
	if err := validateAllowedKeys(network, map[string]struct{}{"allow_endpoints": {}, "deny_endpoints": {}}, "network"); err != nil {
		return err
	}
	for _, key := range []string{"allow_endpoints", "deny_endpoints"} {
		value, ok := network[key]
		if !ok || value == nil {
			continue
		}
		items, ok := value.([]any)
		if !ok {
			return fmt.Errorf("network.%s must be an array of host:port strings", key)
		}
		for _, item := range items {
			endpoint, ok := item.(string)
			if !ok {
				return fmt.Errorf("network.%s must be an array of host:port strings; found non-string element: %v", key, item)
			}
			if err := injectionpolicy.ValidateEgressEndpoint(endpoint, "network."+key); err != nil {
				return err
			}
		}
	}
	return nil
}

func credentialKeyUnion(scoped map[string]map[string]struct{}, shared map[string]struct{}) map[string]struct{} {
	out := make(map[string]struct{})
	for key := range shared {
		out[key] = struct{}{}
	}
	for _, keys := range scoped {
		for key := range keys {
			out[key] = struct{}{}
		}
	}
	return out
}

// PolicySource is an alias for injectionpolicy.PolicySource — the
// canonical cross-package type.  Kept exported here for callers that
// have always imported it as authresolve.PolicySource.
type PolicySource = injectionpolicy.PolicySource

type config struct {
	policyPath     string
	agent          string
	mode           string
	resolutionMode string
	outputPolicy   string
	outputMetadata string
	outputRoot     string
}

// Run executes the resolve-credential-sources workflow.  Diagnostics
// land on stderr exactly as the bash predecessor wrote them, and the
// returned error is either nil or a *cliexit.ExitCodeError carrying
// the bash exit-code contract.  The hostutil wrapper recovers Code via
// errors.As and propagates it to os.Exit, keeping a single typed
// channel for every translated bash main in this repo.
func Run(args []string, stdout, stderr io.Writer) error {
	cfg, err := parseArgs(args)
	if err != nil {
		fmt.Fprintln(stderr, err)
		return &cliexit.ExitCodeError{Code: 2}
	}
	if err := run(cfg); err != nil {
		fmt.Fprintln(stderr, err)
		return &cliexit.ExitCodeError{Code: 1}
	}
	return nil
}

func parseArgs(args []string) (config, error) {
	var cfg config
	for i := 0; i < len(args); i++ {
		arg := args[i]
		if !strings.HasPrefix(arg, "--") {
			return config{}, fmt.Errorf("unexpected argument: %s", arg)
		}
		if i+1 >= len(args) {
			return config{}, fmt.Errorf("missing value for %s", arg)
		}
		value := args[i+1]
		i++
		switch arg {
		case "--policy":
			cfg.policyPath = value
		case "--agent":
			cfg.agent = value
		case "--mode":
			cfg.mode = value
		case "--resolution-mode":
			cfg.resolutionMode = value
		case "--output-policy":
			cfg.outputPolicy = value
		case "--output-metadata":
			cfg.outputMetadata = value
		case "--output-root":
			cfg.outputRoot = value
		default:
			return config{}, fmt.Errorf("unsupported flag: %s", arg)
		}
	}

	if cfg.policyPath == "" {
		return config{}, errors.New("--policy is required")
	}
	if cfg.agent == "" {
		return config{}, errors.New("--agent is required")
	}
	if cfg.mode == "" {
		return config{}, errors.New("--mode is required")
	}
	if cfg.resolutionMode == "" {
		return config{}, errors.New("--resolution-mode is required")
	}
	if cfg.outputPolicy == "" {
		return config{}, errors.New("--output-policy is required")
	}
	if cfg.outputMetadata == "" {
		return config{}, errors.New("--output-metadata is required")
	}
	if cfg.outputRoot == "" {
		return config{}, errors.New("--output-root is required")
	}
	if _, ok := supportedAgents[cfg.agent]; !ok {
		return config{}, fmt.Errorf("unsupported value for --agent: %s", cfg.agent)
	}
	if _, ok := supportedModes[cfg.mode]; !ok {
		return config{}, fmt.Errorf("unsupported value for --mode: %s", cfg.mode)
	}
	if cfg.resolutionMode != "launch" && cfg.resolutionMode != "metadata" {
		return config{}, fmt.Errorf("unsupported value for --resolution-mode: %s", cfg.resolutionMode)
	}
	return cfg, nil
}

func run(cfg config) error {
	policyPath, err := resolvePath(cfg.policyPath)
	if err != nil {
		return err
	}
	outputPolicyPath, err := resolvePath(cfg.outputPolicy)
	if err != nil {
		return err
	}
	outputMetadataPath, err := resolvePath(cfg.outputMetadata)
	if err != nil {
		return err
	}
	outputRoot, err := resolvePath(cfg.outputRoot)
	if err != nil {
		return err
	}
	if err := os.MkdirAll(outputRoot, 0o700); err != nil {
		return err
	}
	if err := os.Chmod(outputRoot, 0o700); err != nil {
		return err
	}
	outputPolicyRel, err := rootio.RelativePathWithin(outputRoot, outputPolicyPath, "--output-policy")
	if err != nil {
		return err
	}
	outputMetadataRel, err := rootio.RelativePathWithin(outputRoot, outputMetadataPath, "--output-metadata")
	if err != nil {
		return err
	}
	outputRootFS, err := os.OpenRoot(outputRoot)
	if err != nil {
		return err
	}
	defer outputRootFS.Close()

	policy, policySources, err := loadPolicyBundle(policyPath)
	if err != nil {
		return err
	}
	policy = rebasePolicyFragment(policy, filepath.Dir(policyPath))
	// Validate [network] fail-closed before resolving any credential; otherwise
	// an invalid [network] (on the resolver whitelist) could stage credentials
	// and write a resolved policy before the later render rejects it.
	if err := validateNetworkPolicy(policy); err != nil {
		return err
	}
	credentials, ok := policy["credentials"]
	if credentials == nil {
		credentials = map[string]any{}
		ok = true
	}
	if !ok {
		return errors.New("credentials must be a TOML table")
	}
	credentialTable, ok := credentials.(map[string]any)
	if !ok {
		return errors.New("credentials must be a TOML table")
	}

	relevantKeys := selectedCredentialKeys(cfg.agent)
	metadata := map[string]any{
		"policy_entrypoint":            logicalPolicyPath(policyPath, filepath.Dir(policyPath)),
		"policy_sources":               policySources,
		"credential_input_kinds":       map[string]string{},
		"credential_resolvers":         map[string]string{},
		"credential_materialization":   map[string]string{},
		"credential_resolution_states": map[string]string{},
		"provider_auth_ready_states":   map[string]string{},
		"shared_auth_ready_states":     map[string]string{},
	}

	for _, key := range relevantKeys {
		raw, exists := credentialTable[key]
		if !exists || raw == nil {
			continue
		}

		rawMap, isMap := raw.(map[string]any)
		if !isMap {
			if _, shared := sharedCredentialKeys[key]; shared {
				return fmt.Errorf("credentials.%s.providers is required so shared GitHub credentials stay least-privilege", key)
			}
			source, err := validateSourcePath(raw, "credentials."+key, cwd())
			if err != nil {
				return err
			}
			if _, err := requireSecretFile(source, "credentials."+key); err != nil {
				return err
			}
			metadata["credential_input_kinds"].(map[string]string)[key] = "source"
			metadata["credential_resolution_states"].(map[string]string)[key] = "source"
			recordAuthReadyState(metadata, key, "ready")
			continue
		}

		if err := validateAllowedKeys(rawMap, CredentialEntryKeys, "credentials."+key); err != nil {
			return err
		}
		resolverName, _ := rawMap["resolver"].(string)
		providers := rawMap["providers"]
		modes := rawMap["modes"]
		if _, shared := sharedCredentialKeys[key]; shared && providers == nil {
			return fmt.Errorf("credentials.%s.providers is required so shared GitHub credentials stay least-privilege", key)
		}

		ok, err := selectedFor(providers, cfg.agent, "credentials."+key+".providers", supportedAgents)
		if err != nil {
			return err
		}
		if !ok {
			recordAuthReadyState(metadata, key, "filtered-provider")
			if resolverName != "" {
				delete(credentialTable, key)
			}
			continue
		}
		ok, err = selectedFor(modes, cfg.mode, "credentials."+key+".modes", supportedModes)
		if err != nil {
			return err
		}
		if !ok {
			recordAuthReadyState(metadata, key, "filtered-mode")
			if resolverName != "" {
				delete(credentialTable, key)
			}
			continue
		}

		_, hasSource := rawMap["source"]
		if hasSource && resolverName != "" {
			return fmt.Errorf("credentials.%s must not declare both source and resolver", key)
		}
		if !hasSource && resolverName == "" {
			return fmt.Errorf("credentials.%s must declare source or resolver", key)
		}
		if resolverName == "" {
			source, err := validateSourcePath(rawMap["source"], "credentials."+key, cwd())
			if err != nil {
				return err
			}
			if _, err := requireSecretFile(source, "credentials."+key); err != nil {
				return err
			}
			metadata["credential_input_kinds"].(map[string]string)[key] = "source"
			metadata["credential_resolution_states"].(map[string]string)[key] = "source"
			recordAuthReadyState(metadata, key, "ready")
			continue
		}

		resolverSet, ok := allowedResolvers[key]
		if !ok {
			return fmt.Errorf("credentials.%s.resolver is unsupported: %s", key, resolverName)
		}
		if _, ok := resolverSet[resolverName]; !ok {
			return fmt.Errorf("credentials.%s.resolver is unsupported: %s", key, resolverName)
		}

		materialization := "ephemeral"
		if rawMaterialization, ok := rawMap["materialization"].(string); ok && rawMaterialization != "" {
			materialization = rawMaterialization
		}
		if _, ok := supportedMaterialization[materialization]; !ok {
			return fmt.Errorf(
				"credentials.%s.materialization must be one of: ephemeral, persistent",
				key,
			)
		}
		if materialization != "ephemeral" {
			return fmt.Errorf(
				"credentials.%s.materialization must stay ephemeral for resolver-backed auth",
				key,
			)
		}

		relativeDestination := filepath.Join("resolved", "credentials", key+".json")
		destination := filepath.Join(outputRoot, relativeDestination)
		state, err := resolveCredential(key, resolverName, outputRootFS, relativeDestination, cfg.resolutionMode)
		if err != nil {
			return err
		}

		rewritten := make(map[string]any, len(rawMap))
		for k, v := range rawMap {
			rewritten[k] = v
		}
		delete(rewritten, "resolver")
		delete(rewritten, "materialization")
		rewritten["source"] = destination
		credentialTable[key] = rewritten

		metadata["credential_input_kinds"].(map[string]string)[key] = "resolver"
		metadata["credential_resolvers"].(map[string]string)[key] = resolverName
		metadata["credential_materialization"].(map[string]string)[key] = materialization
		metadata["credential_resolution_states"].(map[string]string)[key] = state
		recordAuthReadyState(metadata, key, authReadyStateForResolution(state))
	}

	renderedPolicy, err := renderPolicyTOML(policy)
	if err != nil {
		return err
	}
	if err := writeAtomicText(outputRootFS, outputPolicyRel, renderedPolicy); err != nil {
		return err
	}
	metadataJSON, err := json.MarshalIndent(metadata, "", "  ")
	if err != nil {
		return err
	}
	if err := writeAtomicText(outputRootFS, outputMetadataRel, string(metadataJSON)+"\n"); err != nil {
		return err
	}
	return nil
}

func recordAuthReadyState(metadata map[string]any, key, state string) {
	if _, ok := sharedCredentialKeys[key]; ok {
		metadata["shared_auth_ready_states"].(map[string]string)[key] = state
		return
	}
	metadata["provider_auth_ready_states"].(map[string]string)[key] = state
}

func authReadyStateForResolution(state string) string {
	switch state {
	case "resolved", "source", "host-source":
		return "ready"
	default:
		return state
	}
}

func selectedCredentialKeys(agent string) []string {
	keys := make([]string, 0, len(sharedCredentialKeys)+len(agentScopedCredentialKeys[agent]))
	if adapters.SharedCredentialsApplyToAgent(agent) {
		for key := range sharedCredentialKeys {
			keys = append(keys, key)
		}
	}
	for key := range agentScopedCredentialKeys[agent] {
		keys = append(keys, key)
	}
	slices.Sort(keys)
	return keys
}
